### ISSUE TITLE
Llama3 model family - list of required ops for blackhole

### ISSUE THREAD

**Issue by mtairum** [tenstorrent/tt-metal#16013]
## Additional issues
Started tracking full end-to-end Llama3 on Blackhole: https://github.com/tenstorrent/tt-metal/issues/18135.

- DRAM-sharded matmul issue: https://github.com/tenstorrent/tt-metal/issues/18144

## Description
This issue lists the ops required for the Llama8B model (and the rest of the llama3 model family).

Looking at the current list of supported Blackhole ops, the following seem to be the ops we'll require to properly support Llama3 family in blackhole:

- ✅ `ttnn.embedding` 
- ✅ `ttnn.transpose`
- ✅ `ttnn.slice`
- ✅ `ttnn.interleaved_to_sharded`
- ✅ `ttnn.sharded_to_interleaved`
- ✅ `ttnn.layer_norm`
  - #16721 
- ✅`ttnn.rms_norm` (see comments below)
  - #16721 
  - `ttnn.rms_norm_pre_all_gather` - Requires 4+ chips [outside scope of this ticket]
  - `ttnn.rms_norm_post_all_gather` - Requires 4+ chips [outside scope of this ticket]
- ✅ `ttnn.matmul`
- 🟨  [**not blocking Llama3**] `ttnn.experimental.nlp_create_qkv_heads_decode`
  - #16667
- ✅ `tnn.experimental.rotary_embedding_llama`
- ✅ `ttnn.experimental.paged_update_cache`
  - #16674
- ❌ `ttnn.transformer.scaled_dot_product_attention_decode` (with and without paged_attention flag)
  - #16673
  - Potentially related:
    - https://github.com/tenstorrent/tt-metal/issues/16662
    - https://github.com/tenstorrent/tt-metal/issues/16439
- ✅ `ttnn.experimental.nlp_concat_heads_decode`
- ✅ `ttnn.add`
- ✅ `ttnn.mul`
  - https://github.com/tenstorrent/tt-metal/issues/16662
- ✅ `ttnn.concat` 
- ❌ `ttnn.argmax`
  - #16719 
  - #16720
  - This will be superseeded by Top-P/Top-K op when it gets to main: #16854 

Prefill only ops:
- `ttnn.transformer.scaled_dot_product_attention` (if not using chunks)
  - `ttnn.transformer.chunked_scaled_dot_product_attention` (if using chunks - not in the traces. This is the same op as previous one but with page table and chunk start index)
  - Potentially related issues:
    - https://github.com/tenstorrent/tt-metal/issues/16662
    - https://github.com/tenstorrent/tt-metal/issues/16439
- ✅ `ttnn.experimental.nlp_create_qkv_heads`
- ✅ `ttnn.experimental.paged_fill_cache`
- ✅ `ttnn.experimental.nlp_concat_heads`

Below are the graph trace and the perf trace with extra info on the ops (including memory configs and shapes).

## Updated traces [14 Jan 2025]

Please use these new traces for 1B, 8B and 70B llama3 model. These include both prefill and decode and were taken by running the `demo.py` script with 1L, for 10 iterations.

- Llama3-1B: [llama-1b-1L-demo-ops.csv](https://github.com/user-attachments/files/18411407/llama-1b-1L-demo-ops.csv)

- Llama3-8B: [llama-8b-1L-demo-ops.csv](https://github.com/user-attachments/files/18411408/llama-8b-1L-demo-ops.csv)

- Llama3-70B: [llama-70b-1L-demo-ops.csv](https://github.com/user-attachments/files/18411417/llama-70b-1L-demo-ops.csv)
 
## [OLD] Graph Trace
The list of ops was generated with ttnn graph trace:
```
ttnn.graph.begin_graph_capture(ttnn.graph.RunMode.NORMAL)

(...) Llama3 - 8B model run

captured_graph = ttnn.graph.end_graph_capture() # End capturing the graph
ttnn.graph.pretty_print(captured_graph)
ttnn.graph.visualize(captured_graph, file_name="graph.svg")
```

[llama8b-1L-op_graph.txt](https://github.com/user-attachments/files/18129482/llama8b-op_graph.txt)
![Image](https://github.com/user-attachments/assets/d6a47e56-7fb7-4649-b050-780641be34ab)

## Ops Perf report

Generated with tracy, this is the ops perf report, which includes the memory configs and input shapes of the required ops
- [OLD TRACE, USE THE NEW ONE ABOVE] [llama8B-1L-model-ops-perf.csv](https://github.com/user-attachments/files/18129569/llama8B-1L-model-ops-perf.csv)

## Llama3-70B

Additionally, we'll also want support for Llama3-70B ops, which are mostly the same but with different input sizes.
In this section I'll list any new ops separately, and provide the ops perf report.

Additional ops:
- ttnn.all_gather
- ttnn.reduce_scatter

### Ops Perf report
- [OLD TRACE, Please use the new one above] [llama-70b-1L-ops-pers.csv](https://github.com/user-attachments/files/18155395/llama-70b-1L-ops-pers.csv)

**Comment by mtairum** [comment]
@prajaramanTT Are you the right person to tag on this issue?

In the model team we want to understand what's the current op support in blackhole and what's missing for us to support Llama3.

For now, this issue is listing Llama3-8B, which will run on a single device. We want to provide the list of ops + shapes required so those can be added to the ttnn op sweep tests soon. 

Let me know of next steps and please tag other relevant people on this 🙇

**Comment by mtairum** [comment]
FIY @uaydonat

**Comment by mtairum** [comment]
Added ops for Llama3-70B as well. 
It's mostly the same as 8B, but since it's exclusively multichip, has CCL ops: all-gather and reduce scatter.

**Comment by ntarafdar** [comment]
@yugi957 has done great work , most TMs are accounted for and work on BH (embedding, slice, transpose, sharded_to_interleaved) . He will test the remaining ones tomorrow (interleaved_to_sharded, concat)

**Comment by ntarafdar** [comment]
@yugi957 has confirmed all TMs that worked on WH for llama work on BH.

**Comment by mtairum** [comment]
@ntarafdar that's great to hear.

What about other ops, such as the experimental ones (rotary embedding, paged dot product, etc.). Any chance to add these to the BH sweeps and test them there?

These will be crucial in supporting transformer-based LLMs in BH.

**Comment by bbradelTT** [comment]
@vsureshTT will look at
- ttnn.layer_norm
- ttnn.arg_max

I created https://github.com/tenstorrent/tt-metal/issues/16525 to track that effort.

**Comment by uaydonat** [comment]
@cmaryanTT mentioned that someone will be assigned to custom ops (e.g. nlp_create_qkv_heads_decode, rotary_embedding_llama, paged_scaled_dot_product_attention_decode, etc.).

**Comment by cmaryanTT** [comment]
@ntarafdar will be assigning someone in his group to look at the custom ops

**Comment by cmaryanTT** [comment]
@yugi957 and @amorrisonTT will be looking at the custom transformer ops. ETA Monday.
ttnn.experimental.nlp_create_qkv_heads_decode
tnn.experimental.rotary_embedding_llama
ttnn.experimental.paged_update_cache
ttnn.transformer.paged_scaled_dot_product_attention_decode
ttnn.experimental.nlp_concat_heads_decode

**Comment by amorrisonTT** [comment]
These traces don't appear to contain, 

`ttnn.transformer.paged_scaled_dot_product_attention_decode` (although there is `scaled_dot_product_attention_decode`)
`ttnn.add`
`ttnn.mul`

Is this expected?

**Comment by cmaryanTT** [comment]
@amorrisonTT add and mult show up as "BinaryDeviceOperation" with the MULT or ADD type.

**Comment by cmaryanTT** [comment]
I think "paged" in the other op is just a typo

**Comment by amorrisonTT** [comment]
> [@amorrisonTT](https://github.com/amorrisonTT) add and mult show up as "BinaryDeviceOperation" with the MULT or ADD type.

Thanks!

**Comment by bbradelTT** [comment]
@mtairum the only reference to argmax is on row 1960 of [llama-70b-1L-ops-pers.csv](https://github.com/user-attachments/files/18155395/llama-70b-1L-ops-pers.csv) and it's for torch argmax and does not specify any inputs.

What do we need to check for ttnn.argmax?

**Comment by uaydonat** [comment]
@amorrisonTT I think `paged_scaled_dot_product_attention_decode` is the right op. It is contained in 8B trace, not sure why it is not in 70B. @mtairum  maybe you did not have Salar's vllm changes?

Also, we run`argmax` on device for batch=1, and on host if batch>1. I am guessing the traces are for batch=32, that's why you only see torch.argmax. It would be good verify ttnn.argmax since we will ultimately extend it for batch=32 as well, but it is lower priority.

**Comment by mtairum** [comment]
> @amorrisonTT add and mult show up as "BinaryDeviceOperation" with the MULT or ADD type.

This is correct.

> the only reference to argmax is on row 1960 of [llama-70b-1L-ops-pers.csv](https://github.com/user-attachments/files/18155395/llama-70b-1L-ops-pers.csv) and it's for torch argmax and does not specify any inputs.

Re: argmax, it's what Utku mentioned. 
We use `ttnn.argmax` with multicore if batch_size==1 (otherwise, it's single core). and the inputs are of shape `[1, 1, 32, 128256]` for either model.

``` tt_out_tok = ttnn.argmax(tt_out_rm, dim=3, use_multicore=False if batch_size > 1 else True, output_tensor=tt_out_tok)```

and from a new tracy I just generated, the line for argmax:

```
ArgMax,tt_dnn_device,18972,1,{'dim': '3'; 'output_dtype': 'DataType::UINT32'; 'output_mem_config': 'MemoryConfig(memory_layout=TensorMemoryLayout::INTERLEAVED;buffer_type=BufferType::L1;shard_spec=std::nullopt)'; 'use_multicore': 'true'},,47,,52725463868,53033342829,307878961,656089241697,656089887765,303135332,646068,644470,,644470,,,,,,,1,1,32,128256,ROW_MAJOR,BFLOAT16,DEV_1_L1_INTERLEAVED,1,1,1,32,ROW_MAJOR,UINT32,DEV_1_DRAM_INTERLEAVED,,,,,,,,,,,,,,,,,,,,,,1,1,1,32,ROW_MAJOR,UINT32,DEV_1_DRAM_INTERLEAVED,,,,,,,,,,,,,,,,,[],[],['ttnn/cpp/ttnn/operations/reduction/argmax/device/kernels/reader_argmax_interleaved_multicore.cpp'],['reader_argmax_interleaved_multicore/7829460339933776415/'],0,1344,0,0,0,0,1,1,1,[8208384.0],[128.0],307448723,3593
```

>  I think paged_scaled_dot_product_attention_decode is the right op. It is contained in 8B trace, not sure why it is not in 70B. @mtairum maybe you did not have Salar's vllm changes?

We care about both:
- `scaled_dot_product_attention_decode`
- `paged_scaled_dot_product_attention_decode`

It's weird to me why the 70B trace above is not using the paged implementation, I'm pretty sure I run the same config on both. In either case we support both versions, and both have the same input shapes.

**Comment by amorrisonTT** [comment]
>  > I think paged_scaled_dot_product_attention_decode is the right op. It is contained in 8B trace, not sure why it is not in 70B. [@mtairum](https://github.com/mtairum) maybe you did not have Salar's vllm changes?
> 
> We care about both:
> 
> * `scaled_dot_product_attention_decode`
> * `paged_scaled_dot_product_attention_decode`
> 
> It's weird to me why the 70B trace above is not using the paged implementation, I'm pretty sure I run the same config on both. In either case we support both versions, and both have the same input shapes.

I didn't see `paged_scaled_dot_product_attention_decode` in either attached trace:

```
PERF_FILES = ["llama8B-1L-model-ops-perf.csv", "llama-70b-1L-ops-pers.csv"]
df = pd.concat([pd.read_csv(f) for f in PERF_FILES])
df["OP CODE"].value_counts()

OP CODE
Matmul                                 249
InterleavedToShardedDeviceOperation    147
LayerNorm                              147
BinaryDeviceOperation                  147
ReshardDeviceOperation                 145
AllGather                              144
ShardedToInterleavedDeviceOperation    101
Embeddings                              98
Transpose                               98
SliceDeviceOperation                    98
RotaryEmbeddingLlama                    98
PagedUpdateCacheDeviceOperation         98
(torch) __getitem__                     55
NLPCreateHeadsDecodeDeviceOperation     49
ScaledDotProductAttentionDecode         49
NLPConcatHeadsDecodeDeviceOperation     49
ReduceScatter                           48
AllGatherMatmul                         48
(torch) cat                             36
(torch) reshape                         24
(torch) abs                             18
(torch) transpose                       12
(torch) item                            12
(torch) max                             12
(torch) sub                             12
(torch) __get__                          7
(torch) squeeze                          7
(torch) div                              6
(torch) allclose                         6
(torch) permute                          6
ConcatDeviceOperation                    1
(torch) argmax                           1
(torch) embedding                        1
(torch) tolist                           1
Name: count, dtype: int64
```

**Comment by uaydonat** [comment]
Hmm, for 8B, the op graph has the `paged_scaled_dot_product_attention_decode` but not the trace. 

It might be some wrong naming, because `ScaledDotProductAttentionDecode` has `paged_attention: true` in its arguments for both 8B and 70B.

**Comment by amorrisonTT** [comment]
`ttnn.experimental.nlp_create_qkv_heads_decode` failing with,
```
PCC value: Max ATOL Delta: 0.9921875, Max RTOL Delta: 79360.0, PCC: 0.49738503615123153, PCC check failed
2025-01-10 23:28:15.657 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_nlp_create_qkv_heads_decode:run_test_create_head_interleaved:66 - PCC value: Max ATOL Delta: 0.98828125, Max RTOL Delta: 18944.0, PCC: 0.5012214626471699, PCC check failed
2025-01-10 23:28:15.660 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_nlp_create_qkv_heads_decode:run_test_create_head_interleaved:70 - PCC value: Max ATOL Delta: 0.9921875, Max RTOL Delta: 7264.0, PCC: 0.4905412229394511, PCC check failed
```

**Comment by amorrisonTT** [comment]
`ttnn.experimental.paged_update_cache ` failing with:
```
Target number of cores 1024 is greater than total number of available cores 130
```

**Comment by cmaryanTT** [comment]
Per @eyonland ADD works, MULT has a non-determinism issue (https://github.com/tenstorrent/tt-metal/issues/16662)

**Comment by ntarafdar** [comment]
@amorrisonTT when you make the update_cache issue assign it to @cglagovich

**Comment by amorrisonTT** [comment]
ttnn.transformer.scaled_dot_product_attention_decode (with and without paged_attention flag) is failing see #16673

**Comment by vsureshTT** [comment]
@mtairum 
The shapes required for layer norm are

input 0: [1,1,32,8192] input 1: [1,1,256,32]
and
input 0: [1,1,32,4096] input 1: [1,1,128,32]

I linked a truncated version of the spreadsheet below with the layernorm isolated.

[LayerNorms1.csv](https://github.com/user-attachments/files/18402727/LayerNorms1.csv)

**Comment by mtairum** [comment]
I've added new traces and two new prefill only ops:
- `ttnn.transformer.scaled_dot_product_attention` (if not using chunks)
- `ttnn.transformer.chunked_scaled_dot_product_attention`(if using chunks - not in the traces. This is the same op as previous one but with page table and chunk start index)

The chunked version was not used in the trace, but it's basically a variation of the main op. This op is separate than the scaled dot product decode that's already being tested.

**Comment by bbradelTT** [comment]
@vsureshTT ran the tests. The test scenarios failed.

I'll update the description with issue numbers. 

I tried on WH and it seems that the behaviour is the same as on BH, which means that this may not block the model.

**Comment by cglagovichTT** [comment]
`paged_scaled_dot_product_attention_decode` hangs and low PCC issues are likely a combination of issues that `mul` and `matmul_block` are facing. updates are in issue #16673

**Comment by cglagovichTT** [comment]
`nlp_create_qkv_heads_decode` is not blocking Llama on BH. There are issues with the op but not for Llama variants/shapes. updates in issue #16667

**Comment by cglagovichTT** [comment]
`paged_update_cache` is not blocking Llama on BH. I tested with the Llama input shapes and it's passing. Issue #16674

**Comment by mtairum** [comment]
@yugi957 and @amorrisonTT, will you also be looking at the custom prefill-only transformer ops?

They've been added to the traces more recently:

- ttnn.transformer.scaled_dot_product_attention (if not using chunks)
  - ttnn.transformer.chunked_scaled_dot_product_attention (if using chunks - not in the traces. This is the same op as previous one but with page table and chunk start index)
- ttnn.experimental.nlp_create_qkv_heads
- ttnn.experimental.paged_fill_cache
- ttnn.experimental.nlp_concat_heads

**Comment by cglagovichTT** [comment]
fyi for those working on the other `scaled_dot_product_attention` ops, I believe they might be susceptible to #16662 and #16439

**Comment by cmaryanTT** [comment]
@mtairum a couple of the issues for @bbradelTT came up as failing on both WH and BH. Can you confirm if the traces for these ops are correct, and show up identically when running the models on WH and BH:
https://github.com/tenstorrent/tt-metal/issues/16721
https://github.com/tenstorrent/tt-metal/issues/16719
https://github.com/tenstorrent/tt-metal/issues/16720

**Comment by uaydonat** [comment]
@mtairum yes, we also need to provide the prefill traces as well, including the exact shapes and configs we are using (since as Colman said above, the ops might be failing or passing for different shapes)

**Comment by amorrisonTT** [comment]
> [@yugi957](https://github.com/yugi957) and [@amorrisonTT](https://github.com/amorrisonTT), will you also be looking at the custom prefill-only transformer ops?
> 
> They've been added to the traces more recently:
> 
> * ttnn.transformer.scaled_dot_product_attention (if not using chunks)
>   
>   * ttnn.transformer.chunked_scaled_dot_product_attention (if using chunks - not in the traces. This is the same op as previous one but with page table and chunk start index)
> * ttnn.experimental.nlp_create_qkv_heads
> * ttnn.experimental.paged_fill_cache
> * ttnn.experimental.nlp_concat_heads

When your double checking and cross-referencing is confirmed I can take a look at these

**Comment by mtairum** [comment]
The most recent added prefill ops are all passing on BH.
-  `ttnn.experimental.nlp_create_qkv_heads`
- `ttnn.experimental.paged_fill_cache`
- `ttnn.experimental.nlp_concat_heads`

Thank you @cglagovichTT for having a look at these and adding the missing shapes to the tests 🔥

**Comment by mtairum** [comment]
I've updated the Layernorm op in the description. It's actually a RMS Norm. Also updated the relevant issue: https://github.com/tenstorrent/tt-metal/issues/16721

We still require `ttnn.layer_norm`. Existing testing of rms norm passes on WH and BH. On Monday I can check the following shapes:

## RMS_Norm
The other Norm we are interested and the ones shown in the trace are:
- `ttnn.rms_norm`
- distributed rms_norm which is subdivided into 2 ops:
  - `ttnn.rms_norm_pre_all_gather`
  - `rms_norm_post_all_gather`
Reference file: `models/common/rmsnorm.py`

I'll update the main issue with this new information.

The respective test scripts are here:

- rms norm: `tests/tt_eager/python_api_testing/unit_testing/misc/test_rmsnorm.py` 
  - Passes on WH and BH.

- Distributed norm (required for Llama when sequence len > 8K, on multichip) (Always a requirement for TG):
  - `tests/ttnn/unit_tests/operations/test_distributed_layernorm.py` 
  - Sharded version: `tests/ttnn/unit_tests/operations/test_distributed_layernorm_sharded.py`
  - **NOTE** This requires 4 chips to properly test.


### For the normal `rms_norm`

The shapes we're interested in are:

- input shape: [1,1,seqlen,hidden_size]
  - **seqlen**: From 128 to 128k (can be tested in powers of 2)
  - **hidden_size**
    - 1B: 2048
    - 3B: 3072
    - 8/11B: 4096
    - 70B: 8192 -> This is only works in distributed mode! No need to try in rms_norm

- **weight shapes**
  - 1B: [1, 1, 64, 32]
  - 3B: [1, 1, 96, 32]
  - 8/11B: [1, 1, 128, 32]


epsilon=1e-05
program and memory configs are none (should also check sharded, although I haven't double check our settings for that yet).

**Comment by mtairum** [comment]
On the latest sweep update, only SDPA ops are blockers right now. 

Argmax is not needed right now. We currently default to host argmax on Llama.
Additionally, we're waiting for the new sampling op to be merged (already in PR here: https://github.com/tenstorrent/tt-metal/pull/17136) and that new op will substitute the argmax.

Thank you everyone for your contributions so far.

**Comment by cmaryanTT** [comment]
@cglagovichTT to try getting the model working with the MMul workaround.

**Comment by ejouretTT** [comment]
@nvukobratTT please confirm the bisect on the MMul Hang (that makes it go away).

**Comment by nvelickovicTT** [comment]
@ejouretTT I've updated the issue tracking matmul hang (https://github.com/tenstorrent/tt-metal/issues/16439). And asked for people to chime in about prioritization.

Asked @cglagovichTT to try the candidate fix branch in tt-llk-bh repo (https://github.com/tenstorrent/tt-llk-bh/tree/nvelickovic/%2316439_matmul_hang). This branch fixes the test debugged in 16439 and RN50 from @mywoodstock. However, I couldn't repro the SDPA issue.
If it doesn't help with SDPA than it might be another issue or more than one issue. It needs to be debugged separately.

**Comment by cglagovichTT** [comment]
I am unable to repro the [SDPA hang](https://github.com/tenstorrent/tt-metal/issues/16673) on main (703a7a0ec0b1715c4564bd0bc47dd10a930cf9ac) with https://github.com/tenstorrent/tt-llk-bh/tree/nvelickovic/%2316439_matmul_hang

**Comment by uaydonat** [comment]
@cglagovichTT if SDPA is now passing, are we unblocked to run the whole llama8B? Can we try it?

**Comment by cglagovichTT** [comment]
Yes we should be unblocked to try full Llama 8B

**Comment by ejouretTT** [comment]
@cmaryanTT scrub if we have all the variations of Matmul - identify all separate issues. We have good workarounds for Llama 8B and RNET-50. 
@nvelickovicTT to find long term fix. 
@uaydonat to try Llama3.1-8B end to end.

**Comment by ejouretTT** [comment]
Still need to try Llama3.1-8B end to end.

**Comment by mtairum** [comment]
I've started the work on testing full Llama3-8B on blackhole. New issue here: https://github.com/tenstorrent/tt-metal/issues/18135

Linked in the description as well.

**Comment by mtairum** [comment]
Added DRAM-sharded matmul unit test that's failing on Blackhole:

`pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_sharded.py::test_llama_mlp_width_sharded_to_interleaved_pcc_err`

**Comment by mtairum** [comment]
The individual ops required for LLama3 are all done, with a few ongoing issues being tracked at their respective tickets.

Closing this one, as the e2e will be the new issue tracking the LLama3 modules and full model compatibility on Blackhole: https://github.com/tenstorrent/tt-metal/issues/18135
